### ISSUE TITLE
Tagged resources should not re-use check containers from workers that don't have the configured tags

### ISSUE THREAD

**Issue by cirocosta** [concourse/concourse#3002]
## Bug Report

I noticed that it might be possible to a resource that can only be checked in a worker that's tagged with a specific tag, but have that never happen due to global resources taking just `source` in consideration (if I understood correctly).

For instance, consider the case where only `aws`-tagged workers are capable of performing certain actions (like, accessing S3 buckets because they have IAM Roles attached to them):

```yaml
- name: release-from-aws
  tags: [ aws ]
  type: s3
  source: &release-source
    bucket: sample-concourse-s3-resource
    initial_path: binary-1.0.0.tgz
    regexp: binary-(.*).tgz
    region_name: us-east-1
    skip_download: true

- name: release-from-gke
  type: s3
  source: *release-source
```

I'm not sure if this is a *bug* or not - should we consider `tags`?

## Steps to Reproduce

1. have two workers - one tagged, the other not
2. create a pipeline with two resources that share the same config, but differ by having `tags` in one of them
3. verify that the checks always happen in one of the workers (global resources!) - not necessarily in the one with the tag.

## Expected Results

I'm not really sure; eventually, happen over both? Although, I guess that could generate weird results if the worker is not only providing credentials. Maybe this could even end up leaking info somehow?

## Version Info

* Concourse version: master (c862322702eea2fd561b1a04de18ab1c1ef4810b)

**Comment by vito** [comment]
Ugh, this is true.

I really regret supporting IAM roles; they're a perfect example of our 'worker state' anti-pattern, but too many people kept asking for them... 🙄

So is the fundamental problem here that the actual 'credentials' are associated to the machine, and not to the 'source'? That would mean that someone could just configure the same source without having access to it and then obtain the version history regardless of their worker's IAM configuration.

They at least couldn't fetch the bits, and couldn't access any cached bits, since they'd only be cached on workers that have IAM access to fetch them in the first place. But it is a scary information leak if they can see the version listing without credentials.

**Comment by vito** [comment]
In the short-term it may be wise to at least disable resource version history sharing for the s3 resource (and any others that use IAM roles). Thankfully we just implemented support for that in #2908.

We could also make this functionality opt-in by default until we're sure it's safe.

**Comment by vito** [comment]
There are two aspects to this:

1. Not re-using an untagged container for a tagged resource's check
1. Not sharing version history for tagged resources (assuming they may be tagged due to IAM roles or something that made the pipeline author decide to tag the resource)

The second aspect is easy with #2908 and #3013 done (we can just force it to be unique history if the resource has tags), but the first aspect is hard. I propose we do #3079 instead which will remove re-using of check containers entirely.

I think that as long as we finish #2951 prior to 5.0 both of these changes can be deferred until after 5.0 is out, now that we've feature-flagged global resources. I'm OK with this being a known limitation, as anyone using IAM roles should just leave global resources disabled for now anyway.

**Comment by vito** [comment]
We ended up doing this after all in #3133 - reopening (it'll be closed again once the PR is merged).
